Delegate `Rational` invariant to `PNonZero` + optimizations

TotallyNotChase opened the pull request:
The "fail on zero" invariant should be forced in the `PlutusType` instance - otherwise `pcon` is a trap for users.

The golden changes are purely within dev mode and only exist due to a slight change in the error trace.

Geometer1729 reviewed (dismissed):
LGTM

L-as commented:
I'm not sure this is the best solution. Shouldn't we just not export the constructors, or otherwise make the divisor (or dividend?) a Positive?

TotallyNotChase commented:
I think `pcon` is a pretty good notion of smart constructor that already exists. Unless we require `pcon` to never error - I don't see a reason to not utilize it for this purpose.

We could indeed just not export the constructor but then the `PlutusType` instance becomes a bit useless - unless you go on to introduce "safe" helper functions to help with `pmatch`'s handler and `pcon`'s argument, which seems like unnecessary extra work.

L-as commented:
The `PlutusType` instance would still be used internally.

in general, in Haskell, people expect constructors to be total. I don't think we should add partial constructors.

TotallyNotChase commented:
Not wanting `pcon` to be partial is a valid concern - though I don't know how applicable it is for Plutarch usecases. Do you think we should unexport the constructor and just have 2 functions for building them (one returning `PMaybe`, the other being partial)?

And for deconstruction, I suppose the `pnominator` and `pdenominator` will do? I feel like this might be a blocker for achieving maximal performance in userland (using `pmatch` directly in a utility function that is then hoisted will be more efficient) - but oh well.

L-as commented:
Well that's why I suggested Positive.

TotallyNotChase commented:
I assume you mean `NonZero`? Hmm, I could try that and see what it entails.

TotallyNotChase commented on Plutarch/Rational.hs:
```diff
@@ -21,121 +22,131 @@ import Plutarch (
+
+instance PlutusType PRational where
+  type PInner PRational b = (PInteger :--> PNonZero :--> b) :--> b
+  pcon' (PRational x y') = plet y' $ \y -> plam $ \f -> f # x # y
```
I had to make sure the denominator was evaluated prior to building the lambda. Otherwise, you could make a `n / 0` and the error would only happen upon actually matching on the rational.

It's inefficient sometimes - so there's also `pratLazy` below.

TotallyNotChase replied:
For this PR, should I just remove the `PlutusType` instance and have 2 construction functions? (one without the invariant check, and one with it)

L-as commented:
Shouldn't it be `Positive` and not `NonZero`? I don't see why it would be negative.

TotallyNotChase commented:
> Shouldn't it be `Positive` and not `NonZero`? I don't see why it would be negative.

The original invariant was `NonZero` so far so I preserved that behavior. I don't see anything technically wrong with allowing negative denominators but we can definitely change it to `Positive` if necessary.

L-as commented on Plutarch/NonZero.hs:
```diff
@@ -0,0 +1,44 @@
+      $ pcon . PJust . pcon $ PNonZero i
+
+-- | Partial version of 'pnonZero'. Errors if argument is zero.
+pnonZero' :: Term s (PInteger :--> PNonZero)
```
We have the not-so-followed practice of using `ptry` for partial functions.

TotallyNotChase replied:
updated

L-as reviewed (changes requested):
We need to find a general solution to this.

TotallyNotChase commented:
I removed the constructor export and kept the `PlutusType` instance "normal" (no forced invariant check), and added supplementary functions for construction and deconstruction

L-as commented on Plutarch/Reducible.hs:
```diff
@@ -15,3 +15,8 @@ instance Reducible a => Reducible (Const a b) where
 instance Reducible a => Reducible (Identity a) where
   type Reduce (Identity a) = Reduce a
+
+newtype Flip f a b = Flip (f b a)
```
I didn't do this on purpose. `Plutarch.Reducible` shouldn't export any helper types, since they're very often ad-hoc, and aren't relevant to the end-user anyway. They're completely internal and are not part of the interface of Plutarch. The `Reducible` instance is generically derived on my local machine, I'll push that later.

TotallyNotChase replied:
It seems weird to have to copy/paste the same block of boilerplate code in 3 separate files in the same project. Though derivation helpers would indeed make it a bit better.

L-as replied:
it would only be one line then

L-as commented on Plutarch/Rational.hs:
```diff
@@ -38,106 +41,128 @@ import Plutarch.Builtin (
+As such, the constructor is not exported and users are expected to use 'prational', 'prationalLazy', and
+'punrational' instead.
+-}
+instance PlutusType PRational where
```
This should not have a manual implementation

L-as commented on Plutarch/Rational.hs:
```diff
@@ -38,106 +41,128 @@ import Plutarch.Builtin (
+
+i.e, if the denominator was zero, it'll error only after 'pmatch'ing on the resulting rational.
+-}
+prationalLazy :: Term s PInteger -> Term s PNonZero -> Term s PRational
```
This isn't lazy anymore

TotallyNotChase replied:
removed, will figure out optimizations later as you said

L-as commented on Plutarch/Rational.hs:
```diff
@@ -38,106 +41,128 @@ import Plutarch.Builtin (
+          PRational a b <- tcont $ pmatch x
+          let res :: Term _ (PBuiltinList (PAsData PInteger))
+              res = pcons # pdata a #$ pcons # pdata (pto b) #$ pnil
+          pure $ punsafeCoerce $ pdata res
```
Why are we using `punsafeCoerce` here

TotallyNotChase replied:
updated to use `pdataImpl` now

L-as commented on Plutarch/Rational.hs:
```diff
@@ -169,60 +194,48 @@ instance Num (Term s PRational) where
-          pmatch x $ \(PRational xn xd) ->
-            pfailOnZero # xn # (pcon (PRational xd xn))
+          pmatch x $ \(PRational xn xd) -> plet (ptryNonZero # xn) $ \denm ->
+            prationalLazy (pto xd) denm
```
Bad code?

L-as commented:
@TotallyNotChase Can you make this PR up to date?

TotallyNotChase commented:
FYI this isn't ready yet - just syncing with staging.

TotallyNotChase commented on Plutarch/NonZero.hs:
```diff
@@ -0,0 +1,77 @@
+instance PNum PNonZero where
+  x #- y = ptryNonZero #$ pto x #- pto y
+
+  pfromInteger 0 = error "pnonZero.pfromInteger: encountered 0"
```
This currently errors at Plutarch compile time - I feel like there should be a way to produce `Term`s that are just descriptive `CompilerError`s (raised by `compile` - eventually).

L-as replied:
pthrow

TotallyNotChase replied:
updated

TotallyNotChase commented on Plutarch/NonZero.hs:
```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.NonZero (PNonZero, pnonZero, ptryNonZero) where
```
I assume the current convention to enforce invariants for types is to make sure the constructor is not exported? You can't have manual invariant checks in `PlutusType` anymore.

If so, `PValue`'s constructor needs to be unexported.

L-as replied:
yep

TotallyNotChase commented:
I'm trying to figure out optimization stops - where invariant checks should be skipped.

L-as commented:
Just put them in everywhere, then we'll make it efficient later.

TotallyNotChase commented:
Alright, will do optimizations later then

TotallyNotChase commented:
> Hmm, why did the `PNonZero` -> `PPositive` change make the benchmarks perform worse? There wasn't really any changes other than `#==` becoming `#<=` - should investigate.

looks like the benchmarks are performed on scripts compiled with tracing mode on - that's not correct.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -44,120 +47,142 @@ import Plutarch.Builtin (
+  = PRational (Term s PInteger) (Term s PPositive)
   deriving stock (Generic)
-  deriving anyclass (PlutusType)
+  deriving anyclass (PlutusType, PEq)
```
Why are we deriving `PEq`? This doesn't seem correct.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -44,120 +47,142 @@ import Plutarch.Builtin (
+    pfromRational #$ pcon $
+      PRational
+        (pconstant $ numerator x)
+        (punsafeDowncast . pconstant $ denominator x)
```
Are we sure `denominator` is always positive?

TotallyNotChase replied:
[it should be](https://hackage.haskell.org/package/base-4.16.2.0/docs/src/GHC.Real.html#%25) - but we can have further haskell level logic here to invert the sign of the numerator if denominator is negative. Probably unnecessary though.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -44,120 +47,142 @@ import Plutarch.Builtin (
-pListToRat :: Term s (PBuiltinList PInteger :--> PRational)
-pListToRat = plam $ \x -> pcon $ PRational (phead # x) (phead #$ ptail # x)
-
-instance PEq PRational where
```
I think we still need this.

TotallyNotChase replied:
oh yes certainly, oversight - added back.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -44,120 +47,142 @@ import Plutarch.Builtin (
+  type PTryFromExcess PData (PAsData PRational) = Flip Term PPositive
+  ptryFrom' opq = runTermCont $ do
+    (_, ld) <- tcont $ ptryFrom @(PAsData (PBuiltinList PData)) opq
+    tcont $ \f -> pif (plength # ld #== 2) (f ()) (ptraceError "ptryFrom(PRational): data list length should be 2")
```
This isn't efficient, but we can leave it for later.

L-as replied:
Add a TODO

TotallyNotChase replied:
I optimized it

L-as commented on Plutarch/Rational.hs:
```diff
@@ -44,120 +47,142 @@ import Plutarch.Builtin (
       # x'
       # y'
 
   x' #- y' =
```
Couldn't we implement this efficiently in terms of `#+`? Add a TODO.

L-as replied:
Smaller code

L-as commented on Plutarch/Rational.hs:
```diff
@@ -192,48 +217,36 @@ instance PFractional PRational where
-          pfailOnZero # xn # pcon (PRational xd xn)
+          pcon $ PRational (pto xd) $ ptryPositive # xn
 
   x' #/ y' =
```
We can implement this in terms of `#*`. Add a TODO please.

TotallyNotChase replied:
added

L-as commented on Plutarch/Rational.hs:
```diff
@@ -192,48 +217,36 @@ instance PFractional PRational where
+    PRational xn xd' <- tcont $ pmatch x
+    xd <- tcont . plet $ pto xd'
+    r <- tcont . plet $ pgcd # xn # xd
+    s <- tcont . plet . signum $ xd
```
...signum? Not `psignum`?

TotallyNotChase replied:
updated - are we preferring the Plutarch typeclass functions in general?

L-as replied:
Good question.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -192,48 +217,36 @@ instance PFractional PRational where
-          plet (pmin # x # y) $ \b ->
-            pgcd' # a # b
+  plam $ \x' y' -> unTermCont $ do
+    x <- tcont . plet $ abs x'
```
`pabs`?

TotallyNotChase replied:
modified

L-as commented on Plutarch/Rational.hs:
```diff
@@ -254,37 +267,42 @@ pmax = phoistAcyclic $ plam $ \a b -> pif (a #<= b) b a
 
 pfromInteger :: Term s (PInteger :--> PRational)
-pfromInteger = phoistAcyclic $ plam $ \n -> pcon $ PRational n 1
+pfromInteger = phoistAcyclic $ plam $ \n -> pcon $ PRational n $ punsafeDowncast 1
```
Why do you have `punsafeDowncast`?

L-as replied:
Literals for `PPositive` should work fine.

L-as reviewed (changes requested):
We mostly just need TODOs, great work, thanks!

L-as reviewed (approved):
LGTM
